### PR TITLE
drt - Variable stop durations

### PR DESCRIPTION
**Pull request by nkuehnel**
New attempt of 
https://github.com/matsim-org/matsim-libs/pull/1809

I tried to incorporate the ideas discussed in the old PR and for now chose "IncrementalStopDurationEstimator" over "AdditionalStopDuration" for naming the second interface.

@sebhoerl @michalmac let me know if this is what you envisioned.

### CODE REVIEW

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java [hunk 1]
```diff
@@ -10,94 +10,97 @@
...
+			return calcDropoffDetourInfoIfPickupToDropoffDetour(toDropoffTT, fromDropoffTT, pickupDetourInfo, insertion.vehicleEntry, drtRequest);
 		}
 
 		InsertionPoint dropoff = insertion.dropoff;
 		VehicleEntry vEntry = insertion.vehicleEntry;
 
 		if (dropoff.newWaypoint.getLink() == dropoff.previousWaypoint.getLink()) {
 			// no detour, no additional stop duration
-			double dropoffTimeLoss = 0;
+			double dropoffTimeLoss = calcAdditionalDropoffStopDurationIfSameLinkAsPrevious(vEntry, insertion.dropoff.index, drtRequest);
```
Comment (michalmac) [inline contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:84]: please update the comment in L 83

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java [hunk 2]
```diff
@@ -109,18 +112,44 @@ private PickupDetourInfo calcPickupDetourInfoIfPickupToDropoffDetour(VehicleEntr
...
+	private double calcAdditionalPickupStopDurationIfSameLinkAsPrevious(VehicleEntry vEntry, int insertionIdx, DrtRequest drtRequest) {
+		DrtStopTask stopTask;
+
+		if (insertionIdx == 0) {
+			var startTask = vEntry.start.task;
+
+			if (startTask.isPresent() && STOP.isBaseTypeOf(startTask.get())) {
+				stopTask = (DrtStopTask) startTask.get();
+			} else {
+				return stopDurationEstimator.calcForPickup(vEntry.vehicle, null, drtRequest);
```
Comment (michalmac) [inline contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:130]: replace with `stopTask = null;` ?

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java [hunk 3]
```diff
@@ -109,18 +112,44 @@ private PickupDetourInfo calcPickupDetourInfoIfPickupToDropoffDetour(VehicleEntr
...
+	private double calcAdditionalDropoffStopDurationIfSameLinkAsPrevious(VehicleEntry vEntry, int insertionIdx, DrtRequest drtRequest) {
+		DrtStopTask stopTask;
+
+		if (insertionIdx == 0) {
+			var startTask = vEntry.start.task;
+
+			if (startTask.isPresent() && STOP.isBaseTypeOf(startTask.get())) {
+				stopTask = (DrtStopTask) startTask.get();
+			} else {
+				return stopDurationEstimator.calcForDropoff(vEntry.vehicle, null, drtRequest);
```
Comment (michalmac) [inline contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:147]: ditto

File: contribs/drt/src/main/java/org/matsim/contrib/drt/schedule/DrtStayTaskEndTimeCalculator.java [hunk 4]
```diff
@@ -30,10 +30,10 @@
 
 public class DrtStayTaskEndTimeCalculator implements ScheduleTimingUpdater.StayTaskEndTimeCalculator {
 
-	final double stopDuration;
+	private StopDurationEstimator stopDurationEstimator;
```
Comment (michalmac) [inline contribs/drt/src/main/java/org/matsim/contrib/drt/schedule/DrtStayTaskEndTimeCalculator.java:33]: `final`?

**Review by michalmac — APPROVED** [review]
Awesome work @nkuehnel and @sebhoerl! 🚀

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java (hunk 2 above)
Comment (michalmac) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:130]: L130 ^

**Comment by michalmac** [comment]
@nkuehnel I think at one point you did some rebasing or cherry picking (instead of merging master into this branch), so my commits got duplicated. Maybe rebase this branch on origin/master. They should simply disappear (as there haven't been a lot of changes recently in drt).

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/DefaultIncrementalStopDurationEstimator.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import org.matsim.contrib.drt.schedule.DrtStopTask;
+import org.matsim.contrib.dvrp.fleet.DvrpVehicle;
+
+/**
+ * Assumes a fixed stop duration for each pickup/dropoff.
+ * @author nkuehnel / MOIA
+ */
+public class DefaultIncrementalStopDurationEstimator implements IncrementalStopDurationEstimator {
+
+	private final double staticStopDuration;
```
[the remark this answers is not in the document]
Comment (nkuehnel) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/DefaultIncrementalStopDurationEstimator.java]: ✅

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java (hunk 1 above)
Comment (nkuehnel) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:83]: ✅

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java (hunk 2 above)
Comment (nkuehnel) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java]: ✅

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java (hunk 3 above)
Comment (nkuehnel) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java]: ✅

File: contribs/drt/src/main/java/org/matsim/contrib/drt/schedule/DrtStayTaskEndTimeCalculator.java (hunk 4 above)
Comment (nkuehnel) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/schedule/DrtStayTaskEndTimeCalculator.java]: ✅

**Comment by nkuehnel** [comment]
@michalmac thanks for your feedback, your suggestions should be factored in now. Also I rebased branches locally and updated the remote branch accordingly.

**Review by sebhoerl — APPROVED** [review]

**Comment by sebhoerl** [comment]
Thanks :)

**Pull request merged**